File: src/test/java/com/koddy/server/member/domain/model/mentor/TimelineTest.java
package com.koddy.server.member.domain.model.mentor;

import com.koddy.server.common.UnitTest;
import com.koddy.server.member.exception.MemberException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.LocalTime;

import static com.koddy.server.member.domain.model.mentor.DayOfWeek.MON;
import static com.koddy.server.member.exception.MemberExceptionCode.SCHEDULE_PERIOD_TIME_MUST_ALIGN;
import static com.koddy.server.member.exception.MemberExceptionCode.SCHEDULE_PERIOD_TIME_MUST_EXISTS;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

@DisplayName("Member/Mentor -> 도메인 [Timeline] 테스트")
class TimelineTest extends UnitTest {
    @Nested
    @DisplayName("Timeline 생성")
    class Construct {
        @Test
        @DisplayName("시간이 입력되지 않으면 Timeline 생성에 실패한다")
        void throwExceptionByTimeNotExists() {
            assertThatThrownBy(() -> Timeline.of(MON, LocalTime.of(9, 0), null))
                    .isInstanceOf(MemberException.class)
                    .hasMessage(SCHEDULE_PERIOD_TIME_MUST_EXISTS.getMessage());
        }

        @Test
        @DisplayName("멘토링 종료시간가 시작시간 이전이면 Timeline 생성에 실패한다")
        void throwExceptionByStartEndNotAlign() {
            final LocalTime startTime = LocalTime.of(10, 0);
            final LocalTime endTime = LocalTime.of(9, 0);

            assertThatThrownBy(() -> Timeline.of(MON, startTime, endTime))
                    .isInstanceOf(MemberException.class)
                    .hasMessage(SCHEDULE_PERIOD_TIME_MUST_ALIGN.getMessage());
        }

        @Test
        @DisplayName("Timeline을 생성한다")
        void construct() {
            final LocalTime startTime = LocalTime.of(9, 0);
            final LocalTime endTime = LocalTime.of(10, 0);

            assertDoesNotThrow(() -> Timeline.of(MON, startTime, endTime));
        }
    }

    @Test
    @DisplayName("주어진 시간이 멘토링 가능 시간에 포함되는지 확인한다")
    void isTimeIncluded() {
        // given
        final LocalTime startTime = LocalTime.of(9, 0);
        final LocalTime endTime = LocalTime.of(10, 0);
        final Timeline timeline = Timeline.of(MON, startTime, endTime);

        // when
        final boolean actual1 = timeline.isTimeIncluded(LocalTime.of(9, 0));
        final boolean actual2 = timeline.isTimeIncluded(LocalTime.of(9, 30));
        final boolean actual3 = timeline.isTimeIncluded(LocalTime.of(10, 0));
        final boolean actual4 = timeline.isTimeIncluded(LocalTime.of(10, 10));

        // then
        assertAll(
                () -> assertThat(actual1).isTrue(),
                () -> assertThat(actual2).isTrue(),
                () -> assertThat(actual3).isTrue(),
                () -> assertThat(actual4).isFalse()
        );
    }
}

File: src/test/kotlin/com/koddy/server/member/domain/model/mentor/TimelineTest.kt
package com.koddy.server.member.domain.model.mentor

import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.toLocalTime
import com.koddy.server.member.exception.MemberException
import com.koddy.server.member.exception.MemberExceptionCode.SCHEDULE_PERIOD_TIME_MUST_ALIGN
import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.FeatureSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.throwable.shouldHaveMessage

@UnitTestKt
@DisplayName("Member/Mentor -> 도메인 [Timeline] 테스트")
internal class TimelineTest : FeatureSpec({
    feature("Timeline 생성") {
        scenario("멘토링 시작 시간이 종료 시간과 같거나 이후면 Timeline을 생성할 수 없다") {
            mapOf(
                "18:00".toLocalTime() to "18:00".toLocalTime(),
                "18:00".toLocalTime() to "17:59".toLocalTime(),
            ).forEach { (startTime, endTime) ->
                shouldThrow<MemberException> {
                    Timeline(
                        dayOfWeek = DayOfWeek.MON,
                        startTime = startTime,
                        endTime = endTime,
                    )
                } shouldHaveMessage SCHEDULE_PERIOD_TIME_MUST_ALIGN.message
            }
        }

        scenario("멘토링 시작 시간이 종료 시간 이전이면 Timeline을 생성할 수 있다") {
            mapOf(
                "18:00".toLocalTime() to "18:01".toLocalTime(),
                "18:00".toLocalTime() to "18:30".toLocalTime(),
            ).forEach { (startTime, endTime) ->
                shouldNotThrowAny {
                    Timeline(
                        dayOfWeek = DayOfWeek.MON,
                        startTime = startTime,
                        endTime = endTime,
                    )
                }
            }
        }
    }

    feature("Timeline's isTimeIncluded") {
        scenario("주어진 시간이 멘토링 가능 시간에 포함되는지 확인한다") {
            val timeline = Timeline(
                dayOfWeek = DayOfWeek.MON,
                startTime = "09:00".toLocalTime(),
                endTime = "13:00".toLocalTime(),
            )

            mapOf(
                "08:30".toLocalTime() to false,
                "08:59".toLocalTime() to false,
                "09:00".toLocalTime() to true,
                "12:50".toLocalTime() to true,
                "13:00".toLocalTime() to true,
                "13:01".toLocalTime() to false,
            ).forEach { (target, result) ->
                timeline.isTimeIncluded(target) shouldBe result
            }
        }
    }
})
